src/screens/swap/swap-form/swap-form.tsx

```diff
@@ -83,23 +75,38 @@ export const SwapForm: FC<SwapFormProps> = ({ inputToken, outputToken }) => {
     };
 
     trackEvent('SWAP_FORM_SUBMIT', AnalyticsEventCategory.FormSubmit, analyticsProperties);
+
+    if (!inputAssets.amount) {
+      return;
+    }
+
     const routingFeeOpParams = await getRoutingFeeTransferParams(
-      bestTradeWithSlippageTolerance,
-      feeAmount,
+      outputAssets.asset,
+      routingFreeAtomic,
       selectedAccount.publicKeyHash,
       tezos
     );
-    const tradeOpParams = await getTradeOpParams(
-      bestTradeWithSlippageTolerance,
-      selectedAccount.publicKeyHash,
+
+    const tradeOpParams = await getRoute3SwapParams(inputAssets.asset, outputAssets.asset, minimumReceivedAmountAtomic);
+
+    if (!tradeOpParams) {
+      return;
+    }
+
+    const { approve, revoke } = await getTransferPermissions(
```

Comment:
could you move logic with `approve, revoke` opParams into `getRoute3SwapParams` fn?